// unocss-preset-weapp/dist/rules.cjs
'use strict';

const _default = require('./shared/unocss-preset-weapp.82c68b51.cjs');
const transform = require('./shared/unocss-preset-weapp.7ca86f21.cjs');
require('./shared/unocss-preset-weapp.51ea3303.cjs');
require('@unocss/core');
require('unplugin-transform-class/utils');



exports.alignments = _default.alignments;
exports.animations = _default.animations;
exports.appearance = _default.appearance;
exports.aspectRatio = _default.aspectRatio;
exports.backgroundStyles = _default.backgroundStyles;
exports.bgColors = _default.bgColors;
exports.borderStyles = _default.borderStyles;
exports.borders = _default.borders;
exports.boxSizing = _default.boxSizing;
exports.containerParent = _default.containerParent;
exports.cssProperty = _default.cssProperty;
exports.cssVariables = _default.cssVariables;
exports.divides = _default.divides;
exports.flex = _default.flex;
exports.flexGridJustifiesAlignments = _default.flexGridJustifiesAlignments;
exports.floats = _default.floats;
exports.fonts = _default.fonts;
exports.gaps = _default.gaps;
exports.grids = _default.grids;
exports.handlerBorderStyle = _default.handlerBorderStyle;
exports.insets = _default.insets;
exports.justifies = _default.justifies;
exports.margins = _default.margins;
exports.opacity = _default.opacity;
exports.orders = _default.orders;
exports.outline = _default.outline;
exports.overflows = _default.overflows;
exports.paddings = _default.paddings;
exports.placements = _default.placements;
exports.positions = _default.positions;
exports.questionMark = _default.questionMark;
exports.rules = _default.rules;
exports.safeArea = _default.safeArea;
exports.sizes = _default.sizes;
exports.spaces = _default.spaces;
exports.svgUtilities = _default.svgUtilities;
exports.tabSizes = _default.tabSizes;
exports.textAligns = _default.textAligns;
exports.textColors = _default.textColors;
exports.textDecorations = _default.textDecorations;
exports.textIndents = _default.textIndents;
exports.textShadows = _default.textShadows;
exports.textStrokes = _default.textStrokes;
exports.transitions = _default.transitions;
exports.verticalAligns = _default.verticalAligns;
exports.willChange = _default.willChange;
exports.zIndexes = _default.zIndexes;
exports.appearances = transform.appearances;
exports.backgroundBlendModes = transform.backgroundBlendModes;
exports.boxShadows = transform.boxShadows;
exports.boxShadowsBase = transform.boxShadowsBase;
exports.breaks = transform.breaks;
exports.contains = transform.contains;
exports.contentVisibility = transform.contentVisibility;
exports.contents = transform.contents;
exports.cursors = transform.cursors;
exports.displays = transform.displays;
exports.fontSmoothings = transform.fontSmoothings;
exports.fontStyles = transform.fontStyles;
exports.hyphens = transform.hyphens;
exports.isolations = transform.isolations;
exports.mixBlendModes = transform.mixBlendModes;
exports.objectPositions = transform.objectPositions;
exports.pointerEvents = transform.pointerEvents;
exports.resizes = transform.resizes;
exports.ringBase = transform.ringBase;
exports.rings = transform.rings;
exports.screenReadersAccess = transform.screenReadersAccess;
exports.textOverflows = transform.textOverflows;
exports.textTransforms = transform.textTransforms;
exports.transformBase = transform.transformBase;
exports.transforms = transform.transforms;
exports.userSelects = transform.userSelects;
exports.varEmpty = transform.varEmpty;
exports.whitespaces = transform.whitespaces;
exports.writingModes = transform.writingModes;
exports.writingOrientations = transform.writingOrientations;
