// unocss-preset-weapp/dist/transformer.cjs
'use strict';

const utils$1 = require('unplugin-transform-class/utils');
const pluginutils = require('@rollup/pluginutils');
const utils = require('unplugin-attributify-to-class/utils');

function transformerClass(options = { classTags: true }) {
  const idFilter = pluginutils.createFilter(
    options.include || [/\.[jt]sx?$/, /\.vue$/, /\.vue\?vue/],
    options.exclude || [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/]
  );
  const vueFilter = pluginutils.createFilter(
    [/\.vue$/, /\.vue\?vue/]
  );
  return {
    name: "transformer-applet-class",
    idFilter,
    enforce: "pre",
    transform(code, id) {
      let newCode = utils$1.transformCode(code.toString(), options.transformRules);
      if (options.classTags) {
        const classNames = utils$1.getClass(code.toString());
        const injectStr = Array.from(new Set(classNames.map((x) => x[1]).filter((x) => x).flatMap((x) => x.split(" ")))).join(" ");
        if (vueFilter(id))
          newCode = newCode.replace("<template>", `<template>
<!-- ${injectStr} -->
`);
        else
          newCode = `/* ${injectStr} */
${code}`;
      }
      code.overwrite(0, code.original.length, newCode);
    }
  };
}

function transformerWeAttributify(options = {}) {
  const idFilter = pluginutils.createFilter(
    options.include || [/\.vue$/, /\.vue\?vue/],
    options.exclude || [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/]
  );
  const extractor = utils.extractorAttributify(options);
  return {
    name: "transformer-applet-attributify",
    idFilter,
    enforce: "pre",
    transform(code) {
      const newCode = extractor(code.toString());
      code.overwrite(0, code.original.length, newCode);
    }
  };
}

exports.defaultAttributes = utils.defaultAttributes;
exports.defaultIgnoreNonValuedAttributes = utils.defaultIgnoreNonValuedAttributes;
exports.transformerAttributify = transformerWeAttributify;
exports.transformerClass = transformerClass;
