// unocss-preset-weapp/dist/transformer.mjs
import { transformCode, getClass } from 'unplugin-transform-class/utils';
import { createFilter } from '@rollup/pluginutils';
import { extractorAttributify } from 'unplugin-attributify-to-class/utils';
export { defaultAttributes, defaultIgnoreNonValuedAttributes } from 'unplugin-attributify-to-class/utils';

function transformerClass(options = { classTags: true }) {
  const idFilter = createFilter(
    options.include || [/\.[jt]sx?$/, /\.vue$/, /\.vue\?vue/],
    options.exclude || [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/]
  );
  const vueFilter = createFilter(
    [/\.vue$/, /\.vue\?vue/]
  );
  return {
    name: "transformer-applet-class",
    idFilter,
    enforce: "pre",
    transform(code, id) {
      let newCode = transformCode(code.toString(), options.transformRules);
      if (options.classTags) {
        const classNames = getClass(code.toString());
        const injectStr = Array.from(new Set(classNames.map((x) => x[1]).filter((x) => x).flatMap((x) => x.split(" ")))).join(" ");
        if (vueFilter(id))
          newCode = newCode.replace("<template>", `<template>
<!-- ${injectStr} -->
`);
        else
          newCode = `/* ${injectStr} */
${code}`;
      }
      code.overwrite(0, code.original.length, newCode);
    }
  };
}

function transformerWeAttributify(options = {}) {
  const idFilter = createFilter(
    options.include || [/\.vue$/, /\.vue\?vue/],
    options.exclude || [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/]
  );
  const extractor = extractorAttributify(options);
  return {
    name: "transformer-applet-attributify",
    idFilter,
    enforce: "pre",
    transform(code) {
      const newCode = extractor(code.toString());
      code.overwrite(0, code.original.length, newCode);
    }
  };
}

export { transformerWeAttributify as transformerAttributify, transformerClass };
