// unocss-preset-weapp/dist/variants.cjs
'use strict';

const _default = require('./shared/unocss-preset-weapp.b25ecdec.cjs');
require('./shared/unocss-preset-weapp.51ea3303.cjs');
require('@unocss/core');
require('unplugin-transform-class/utils');
require('./shared/unocss-preset-weapp.41085966.cjs');



exports.partClasses = _default.partClasses;
exports.variantAria = _default.variantAria;
exports.variantBreakpoints = _default.variantBreakpoints;
exports.variantColorsMediaOrClass = _default.variantColorsMediaOrClass;
exports.variantCombinators = _default.variantCombinators;
exports.variantContainerQuery = _default.variantContainerQuery;
exports.variantCssLayer = _default.variantCssLayer;
exports.variantCustomMedia = _default.variantCustomMedia;
exports.variantDataAttribute = _default.variantDataAttribute;
exports.variantImportant = _default.variantImportant;
exports.variantInternalLayer = _default.variantInternalLayer;
exports.variantLanguageDirections = _default.variantLanguageDirections;
exports.variantNegative = _default.variantNegative;
exports.variantPrint = _default.variantPrint;
exports.variantPseudoClassFunctions = _default.variantPseudoClassFunctions;
exports.variantPseudoClassesAndElements = _default.variantPseudoClassesAndElements;
exports.variantScope = _default.variantScope;
exports.variantSelector = _default.variantSelector;
exports.variantSpaceAndDivide = _default.variantSpaceAndDivide;
exports.variantSupports = _default.variantSupports;
exports.variantTaggedPseudoClasses = _default.variantTaggedPseudoClasses;
exports.variantVariables = _default.variantVariables;
exports.variants = _default.variants;
