// @alova/mock/dist/alova-mock.esm.js
/**
  * @alova/mock 1.4.1 (https://github.com/alovajs/mock)
  * Copyright 2023 JOU-amjs. All Rights Reserved
  * Licensed under MIT (https://github.com/alovajs/mock/blob/master/LICENSE)
  */

const mockLabel = 'Mock';
const mockLabelColor = '#64E8D6';
const realRequestLabel = 'Realtime';
const realRequestLabelColor = '#999999';
const labelStyle = (bgColor, borderColor = '') => {
    let style = `padding: 2px 6px; background: ${bgColor}; color: white;`;
    if (borderColor) {
        style += `border: solid 1px ${borderColor}`;
    }
    return style;
};
const titleStyle = () => 'color: black; font-size: 12px; font-weight: bolder';
const transform2TableData = (obj) => {
    const tableData = {};
    for (const key in obj) {
        tableData[key] = { value: obj[key] };
    }
    return tableData;
};
const consoleRequestInfo = ({ isMock, url, method, headers, query, data, responseHeaders, response }) => {
    console.groupCollapsed(`%c${isMock ? mockLabel : realRequestLabel}`, labelStyle(isMock ? mockLabelColor : realRequestLabelColor), url);
    console.log('%c[Method]', titleStyle(), method.toUpperCase());
    console.log('%c[Request Headers]', titleStyle());
    console.table(transform2TableData(headers));
    console.log('%c[Query String Parameters]', titleStyle());
    console.table(transform2TableData(query));
    console.log('%c[Request Body]', titleStyle(), data || '');
    if (isMock) {
        if (Object.keys(responseHeaders).length > 0) {
            console.log('%c[Response Headers]', titleStyle());
            console.table(transform2TableData(responseHeaders));
        }
        console.log('%c[Response Body]', titleStyle(), response || '');
    }
    console.groupEnd();
};

const defaultMockResponse = ({ status = 200, statusText = 'ok', body }) => ({
    response: new Response(JSON.stringify(body), {
        status,
        statusText
    }),
    headers: new Headers({})
});
const defaultMockError = (error) => error;

const undefinedValue = undefined, trueValue = true, falseValue = false;
const noop = () => { };
const isFn = (arg) => typeof arg === 'function';
const isNumber = (arg) => typeof arg === 'number' && !isNaN(arg);
const isString = (arg) => typeof arg === 'string';
const parseUrl = (url) => {
    url = /^[^/]*\/\//.test(url) ? url : '//' + url;
    const splitedFullPath = url.split('/').slice(3);
    const query = {};
    let pathContainedParams = splitedFullPath.pop(), pathname = '', hash = '';
    if (pathContainedParams) {
        pathContainedParams = pathContainedParams.replace(/\?[^?#]+/, mat => {
            mat
                .substring(1)
                .split('&')
                .forEach(paramItem => {
                const [key, value] = paramItem.split('=');
                key && (query[key] = value);
            });
            return '';
        });
        pathContainedParams = pathContainedParams.replace(/#[^#]*/, mat => {
            hash = mat;
            return '';
        });
        splitedFullPath.push(pathContainedParams);
        pathname = '/' + splitedFullPath.join('/');
    }
    return {
        pathname,
        query,
        hash
    };
};

function MockRequest({ enable = trueValue, delay = 2000, httpAdapter, mockRequestLogger = consoleRequestInfo, mock, onMockResponse = defaultMockResponse, onMockError = defaultMockError } = { mock: {} }) {
    return (elements, method) => {
        mock = (enable && mock) || {};
        const { url, data, type, headers: requestHeaders } = elements;
        const { pathname, query } = parseUrl(url);
        const params = {};
        const pathnameSplited = pathname.split('/');
        const foundMockDataKeys = Object.keys(mock).filter(key => {
            if (key.startsWith('-')) {
                return falseValue;
            }
            let method = 'GET';
            key = key.replace(/^\[(GET|POST|PUT|DELETE|PATCH|HEAD|OPTIONS|TRACE|CONNECT)\]/i, (_, $1) => {
                method = $1.toUpperCase();
                return '';
            });
            if (method !== type.toUpperCase()) {
                return falseValue;
            }
            const keySplited = key.split('/');
            if (keySplited.length !== pathnameSplited.length) {
                return falseValue;
            }
            for (const i in keySplited) {
                const keySplitedItem = keySplited[i];
                const matchedParamKey = (keySplitedItem.match(/^\{(.*)\}$/) || ['', ''])[1];
                if (!matchedParamKey) {
                    if (keySplitedItem !== pathnameSplited[i]) {
                        return falseValue;
                    }
                }
                else {
                    params[matchedParamKey] = pathnameSplited[i];
                }
            }
            return trueValue;
        });
        let finalKey = foundMockDataKeys.find(key => !/\{.*\}/.test(key));
        finalKey = finalKey || foundMockDataKeys.shift();
        const mockDataRaw = finalKey ? mock[finalKey] : undefinedValue;
        if (mockDataRaw === undefinedValue) {
            if (httpAdapter) {
                isFn(mockRequestLogger) &&
                    mockRequestLogger({
                        isMock: falseValue,
                        url,
                        method: type,
                        params,
                        headers: requestHeaders,
                        query,
                        data: {},
                        responseHeaders: {}
                    });
                return httpAdapter(elements, method);
            }
            else {
                throw new Error(`could not find the httpAdapter which send request.\n[url]${url}`);
            }
        }
        let timer;
        let rejectFn = noop;
        const timeout = method.config.timeout || 0;
        if (timeout > 0) {
            setTimeout(() => {
                rejectFn(new Error('request timeout'));
            }, timeout);
        }
        const resonpsePromise = new Promise((resolve, reject) => {
            rejectFn = reject;
            timer = setTimeout(() => {
                try {
                    const res = isFn(mockDataRaw)
                        ? mockDataRaw({
                            query,
                            params,
                            data,
                            headers: requestHeaders
                        })
                        : mockDataRaw;
                    resolve(new Promise((resolveInner, rejectInner) => {
                        rejectFn = rejectInner;
                        Promise.resolve(res).then(resolveInner).catch(rejectInner);
                    }));
                }
                catch (error) {
                    reject(error);
                }
            }, delay);
        })
            .then(response => {
            let status = 200, statusText = 'ok', responseHeaders = {}, body = undefinedValue;
            if (response === undefinedValue) {
                status = 404;
                statusText = 'api not found';
            }
            else if (isNumber(response.status) && isString(response.statusText)) {
                status = response.status;
                statusText = response.statusText;
                responseHeaders = response.responseHeaders || responseHeaders;
                body = response.body;
            }
            else {
                body = response;
            }
            isFn(mockRequestLogger) &&
                mockRequestLogger({
                    isMock: trueValue,
                    url,
                    method: type,
                    params,
                    headers: requestHeaders,
                    query,
                    data: data || {},
                    responseHeaders,
                    response: body
                });
            return onMockResponse({ status, statusText, responseHeaders, body }, {
                headers: requestHeaders,
                query,
                params,
                data: data || {}
            }, method);
        })
            .catch(error => Promise.reject(onMockError(error, method)));
        return {
            response: () => resonpsePromise.then(({ response }) => response.toString() === '[object Response]' ? response.clone() : response),
            headers: () => resonpsePromise.then(({ headers }) => headers),
            abort: () => {
                clearTimeout(timer);
                rejectFn(new Error('The user abort request'));
            }
        };
    };
}

function createAlovaMockAdapter(mockWrapper, options = { enable: true }) {
    let uniqueMockMap = {};
    mockWrapper
        .filter(({ enable }) => enable)
        .forEach(({ data }) => {
        uniqueMockMap = {
            ...data,
            ...uniqueMockMap
        };
    });
    return MockRequest({
        ...options,
        mock: uniqueMockMap
    });
}

var defineMock = (mock, enable = true) => ({
    enable,
    data: mock
});

export { createAlovaMockAdapter, defineMock };
