// @dcloudio/uni-app-uts/dist/plugins/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.stringifyMap = exports.isVue = exports.genClassName = exports.uvueOutDir = exports.parseImports = exports.ENTRY_FILENAME = void 0;
const path_1 = __importDefault(require("path"));
const es_module_lexer_1 = require("es-module-lexer");
const uni_cli_shared_1 = require("@dcloudio/uni-cli-shared");
const shared_1 = require("@vue/shared");
exports.ENTRY_FILENAME = 'index.uts';
async function parseImports(code) {
    await es_module_lexer_1.init;
    const [imports] = (0, es_module_lexer_1.parse)(code);
    return imports
        .map(({ s, e }) => {
        return `import "${code.slice(s, e)}"`;
    })
        .join('\n');
}
exports.parseImports = parseImports;
function uvueOutDir() {
    return path_1.default.join(process.env.UNI_OUTPUT_DIR, '../.uvue');
}
exports.uvueOutDir = uvueOutDir;
function genClassName(fileName) {
    return ('Gen' +
        (0, shared_1.capitalize)((0, shared_1.camelize)((0, uni_cli_shared_1.removeExt)((0, uni_cli_shared_1.normalizePath)(fileName).replace(/\//g, '-')))));
}
exports.genClassName = genClassName;
function isVue(filename) {
    return filename.endsWith('.vue') || filename.endsWith('.uvue');
}
exports.isVue = isVue;
function stringifyMap(obj) {
    return mapToInitString(objToMap(obj), true);
}
exports.stringifyMap = stringifyMap;
function mapToInitString(map, ts = false) {
    let entries = [];
    for (let [key, value] of map) {
        if (value instanceof Map) {
            entries.push(`["${key}", ${mapToInitString(value, ts)}]`);
        }
        else {
            entries.push(`["${key}", ${JSON.stringify(value)}]`);
        }
    }
    return `new Map${ts ? '<string, any>' : ''}([${entries.join(', ')}])`;
}
function objToMap(obj) {
    const map = new Map();
    for (const key in obj) {
        const value = obj[key];
        if (typeof value === 'object') {
            map.set(key, objToMap(value));
        }
        else {
            map.set(key, value);
        }
    }
    return map;
}
