// @alova/adapter-uniapp/dist/alova-adapter-uniapp.esm.js
/**
  * @alova/adapter-uniapp 1.1.1 (https://alova.js.org)
  * Copyright 2023 JOU-amjs. All Rights Reserved
  * Licensed under MIT (https://github.com/alovajs/adapter-uniapp/blob/master/LICENSE)
  */

import VueHook from 'alova/vue';

const noop = () => { };
const isPlainObject = (arg) => Object.prototype.toString.call(arg) === '[object Object]';

const requestAdapter = (elements, method) => {
    const { url, data, type, headers: header } = elements;
    let taskInstance;
    let onDownload = noop, onUpload = noop;
    const responsePromise = new Promise((resolve, reject) => {
        const { config: adapterConfig } = method;
        const { requestType, timeout } = adapterConfig;
        if (requestType === 'upload') {
            const formData = {};
            const fileData = {};
            if (isPlainObject(data)) {
                Object.keys(data).forEach(key => {
                    if (['name', 'files', 'file', 'filePath'].includes(key)) {
                        fileData[key] = data[key];
                    }
                    else {
                        formData[key] = data[key];
                    }
                });
            }
            const uploadTask = (taskInstance = uni.uploadFile({
                ...adapterConfig,
                ...fileData,
                name: fileData.name,
                url,
                header,
                formData,
                timeout,
                success: res => resolve(res),
                fail: reason => reject(new Error(reason.errMsg)),
                complete: noop
            }));
            onUpload = (handler) => {
                uploadTask.onProgressUpdate(({ totalBytesSent, totalBytesExpectedToSend }) => {
                    handler(totalBytesSent, totalBytesExpectedToSend);
                });
            };
        }
        else if (requestType === 'download') {
            const downloadTask = (taskInstance = uni.downloadFile({
                ...adapterConfig,
                url,
                header,
                timeout,
                success: res => resolve(res),
                fail: reason => reject(new Error(reason.errMsg)),
                complete: noop
            }));
            onDownload = (handler) => {
                downloadTask.onProgressUpdate(({ totalBytesWritten, totalBytesExpectedToWrite }) => {
                    handler(totalBytesWritten, totalBytesExpectedToWrite);
                });
            };
        }
        else {
            taskInstance = uni.request({
                ...adapterConfig,
                url,
                data,
                header,
                method: type,
                timeout,
                success: res => resolve(res),
                fail: reason => reject(new Error(reason.errMsg))
            });
        }
    });
    return {
        response: () => responsePromise,
        headers: () => responsePromise.then(res => res.header || {}),
        abort: () => {
            taskInstance.abort();
        },
        onDownload,
        onUpload
    };
};

var storageAdapter = {
    get(key) {
        return uni.getStorageSync(key);
    },
    set(key, value) {
        uni.setStorageSync(key, value);
    },
    remove(key) {
        uni.removeStorageSync(key);
    }
};

const mockResponseHandler = ({ status, body }, _, currentMethod) => {
    const { requestType } = currentMethod.config;
    const responseHeaders = {};
    if (requestType === 'upload') {
        return {
            response: {
                data: body,
                statusCode: status
            },
            headers: responseHeaders
        };
    }
    else if (requestType === 'download') {
        const isSuccess = status === 200;
        return {
            response: {
                tempFilePath: isSuccess ? body : '',
                statusCode: status
            },
            headers: responseHeaders
        };
    }
    return {
        response: {
            data: body,
            statusCode: status,
            header: responseHeaders,
            cookies: []
        },
        headers: responseHeaders
    };
};

function AdapterUniapp({ mockRequest } = {}) {
    return {
        statesHook: VueHook,
        requestAdapter: mockRequest || requestAdapter,
        storageAdapter
    };
}

export { AdapterUniapp as default, mockResponseHandler as uniappMockResponse, requestAdapter as uniappRequestAdapter, storageAdapter as uniappStorageAdapter };
