// @dcloudio/uni-app-uts/dist/plugins/uvue/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.uniAppUVuePlugin = void 0;
const path_1 = __importDefault(require("path"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const uni_cli_shared_1 = require("@dcloudio/uni-cli-shared");
const descriptorCache_1 = require("./descriptorCache");
const error_1 = require("./error");
const utils_1 = require("../utils");
const script_1 = require("./code/script");
const template_1 = require("./code/template");
const style_1 = require("./code/style");
function resolveAppVue(inputDir) {
    const appUVue = path_1.default.resolve(inputDir, 'app.uvue');
    if (fs_extra_1.default.existsSync(appUVue)) {
        return (0, uni_cli_shared_1.normalizePath)(appUVue);
    }
    return (0, uni_cli_shared_1.normalizePath)(path_1.default.resolve(inputDir, 'App.vue'));
}
function uniAppUVuePlugin() {
    const options = {
        root: process.env.UNI_INPUT_DIR,
        sourceMap: false,
        // eslint-disable-next-line no-restricted-globals
        compiler: require('@vue/compiler-sfc'),
    };
    const appVue = resolveAppVue(process.env.UNI_INPUT_DIR);
    function isAppVue(id) {
        return (0, uni_cli_shared_1.normalizePath)(id) === appVue;
    }
    async function transformMain(code, filename, options, pluginContext) {
        // prev descriptor is only set and used for hmr
        const { descriptor, errors } = (0, descriptorCache_1.createDescriptor)(filename, code, options);
        if (errors.length) {
            errors.forEach((error) => pluginContext.error((0, error_1.createRollupError)(filename, error)));
            return null;
        }
        const isApp = isAppVue(filename);
        const fileName = path_1.default.relative(process.env.UNI_INPUT_DIR, filename);
        const className = (0, utils_1.genClassName)(fileName);
        // 生成 script 文件
        pluginContext.emitFile({
            type: 'asset',
            fileName,
            source: (0, script_1.genScript)(descriptor, { filename: className }) +
                '\n' +
                (0, style_1.genStyle)(descriptor, { filename: fileName, className }) +
                '\n' +
                (!isApp
                    ? (0, template_1.genTemplate)(descriptor, {
                        targetLanguage: process.env.UNI_UVUE_TARGET_LANGUAGE,
                        mode: 'function',
                        filename: className,
                        prefixIdentifiers: true,
                    })
                    : ''),
        });
        let jsCode = '';
        const content = descriptor.script?.content;
        if (content) {
            jsCode += await (0, utils_1.parseImports)(content);
        }
        if (descriptor.styles.length) {
            jsCode += '\n' + (await (0, style_1.genJsStylesCode)(descriptor, pluginContext));
        }
        if (!jsCode) {
            jsCode = 'export default {}';
        }
        return {
            code: jsCode,
        };
    }
    return {
        name: 'uni:app-uvue',
        apply: 'build',
        async resolveId(id) {
            // serve sub-part requests (*?vue) as virtual modules
            if ((0, uni_cli_shared_1.parseVueRequest)(id).query.vue) {
                return id;
            }
        },
        load(id) {
            const { filename, query } = (0, uni_cli_shared_1.parseVueRequest)(id);
            // select corresponding block for sub-part virtual modules
            if (query.vue) {
                if (query.src) {
                    return fs_extra_1.default.readFileSync(filename, 'utf-8');
                }
                const descriptor = (0, descriptorCache_1.getDescriptor)(filename, options);
                let block;
                if (query.type === 'style') {
                    block = descriptor.styles[query.index];
                }
                else if (query.index != null) {
                    block = descriptor.customBlocks[query.index];
                }
                if (block) {
                    return {
                        code: block.content,
                        map: block.map,
                    };
                }
            }
        },
        async transform(code, id) {
            const { filename, query } = (0, uni_cli_shared_1.parseVueRequest)(id);
            if (!(0, utils_1.isVue)(filename)) {
                return;
            }
            if (!query.vue) {
                // main request
                return transformMain(code, filename, options, this);
            }
            else {
                // sub block request
                const descriptor = query.src
                    ? (0, descriptorCache_1.getSrcDescriptor)(filename)
                    : (0, descriptorCache_1.getDescriptor)(filename, options);
                if (query.type === 'style') {
                    return (0, style_1.transformStyle)(code, descriptor, Number(query.index), options, this, filename);
                }
            }
        },
    };
}
exports.uniAppUVuePlugin = uniAppUVuePlugin;
